src/github.com/couchbase/sync_gateway/base/rlimit.go

```diff
@@ -4,25 +4,97 @@ package base
 
 import "syscall"
 
-func SetMaxFileDescriptors(maxFDs uint64) (uint64, error) {
+// Set Max File Descriptor limits
+//
+// Background information:
+//
+// - SG docs
+//   http://developer.couchbase.com/documentation/mobile/1.1.0/develop/guides/sync-gateway/os-level-tuning/max-file-descriptors/index.html
+// - Related SG issues
+//   https://github.com/couchbase/sync_gateway/issues/1083
+// - Hard limit vs Soft limit
+//   http://unix.stackexchange.com/questions/29577/ulimit-difference-between-hard-and-soft-limits
+func SetMaxFileDescriptors(requestedSoftFDLimit uint64) (uint64, error) {
+
+	requiresUpdate, recommendedSoftFDLimit, err := getSoftFDLimit(requestedSoftFDLimit)
+	if err != nil {
+		return 0, err
+	}
+
+	// No call to Setrlimit required, because the requested soft limit is lower than current soft limit
+	if !requiresUpdate {
+		return 0, nil
+	}
+
+	// Get existing limits
 	var limits syscall.Rlimit
 	if err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &limits); err != nil {
-		return maxFDs, err
+		return 0, err
 	}
-	if maxFDs > limits.Max {
-		maxFDs = limits.Max
+
+	// Update the soft limit (but don't bother updating the hard limit, since only root can do that,
+	// and it's assumed that this process is not running as root)
+	limits.Cur = recommendedSoftFDLimit
+	err = syscall.Setrlimit(syscall.RLIMIT_NOFILE, &limits)
+
+	if err == nil {
+		Logf("Configured process to allow %d open file descriptors", recommendedSoftFDLimit)
 	}
-	if limits.Cur == maxFDs {
-		return maxFDs, nil
+
+	return recommendedSoftFDLimit, err
+
+}
+
+func getSoftFDLimit(requestedSoftFDLimit uint64) (requiresUpdate bool, recommendedSoftFDLimit uint64, err error) {
+
+	var limits syscall.Rlimit
+
+	if err := syscall.Getrlimit(syscall.RLIMIT_NOFILE, &limits); err != nil {
+		return false, 0, err
 	}
-	limits.Cur = maxFDs
-	limits.Max = maxFDs
 
-	err := syscall.Setrlimit(syscall.RLIMIT_NOFILE, &limits)
+	requiresUpdate, recommendedSoftFDLimit = getSoftFDLimitWithCurrent(requestedSoftFDLimit, limits.Cur, limits.Max)
 
-	if err == nil {
-		Logf("Configured process to allow %d open file descriptors", maxFDs)
+	return requiresUpdate, recommendedSoftFDLimit, nil
+
+}
+
+// Get the recommended file descriptor settings
+//
+// Given:
+//
+// 1. The max file descriptors requested in the config (or default value)
+// 2. The current ulimit settings
+//
+// return the recommended soft limit for the number of open file descriptors
+// this process can have open.
+//
+// Rules:
+//
+// 1. Only return a value that is HIGHER than the existing soft limit, since
+//    a very small percentage of users will wanting to be lowering the limit
+//    based on this configuration.
+// 1. Only return a value that is LESS-THAN-OR-EQUAL to the existing hard limit
```

Comment:
Oh, I guess I've gotten too accustomed to github flavored markdown.  Good catch.